### PR TITLE
feat: add `ak.merge_union_of_records`

### PR DESCRIPTION
**Pull request by agoose77**
This PR addresses #2182. We already implement some of this machinery in `UnionArray`, which has a `_union_of_optionarrays` method to convert from `option[union[X, Y, Z]]` to `union[option[X], Y, ..., Z]`. It doesn't seem like that is particularly useful in this case. 

Interestingly there is some overlap with `ak._util.union_to_record`.

### Added functions
- `ak.merge_option_of_records`
- `ak.merge_union_of_records`

### CODE REVIEW

File: src/awkward/operations/ak_merge_union_of_records.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+
+def _impl(array, axis, highlevel, behavior):
+    behavior = ak._util.behavior_of(array, behavior=behavior)
+    layout = ak.to_layout(array, allow_record=False)
+
+    def apply_displace_index(layout, backend, **kwargs):
+        if layout.is_record:
+            return layout
+        elif (layout.is_option or layout.is_indexed) and layout.content.is_record:
```
Comment (agoose77) [inline src/awkward/operations/ak_merge_union_of_records.py]: This is doing something different for `is_option`. `merge_option_of_records`

**Comment by agoose77** [comment]
@jpivarski what I'm not certain on here is what should happen for esoteric layouts that have multiple "option-of-record", and "union-of-record" in a single dimension. What should we do here, first wins? It seems like we do that in `ak.drop_none()`, but I don't recall us having a rule.

**Review by jpivarski — APPROVED** [review]
I added some tests which include the original problem in #2182 and a variation of it that has already-optional fields (to verify that the `simplified` classmethod is being used).

I also added tests for the `ak.merge_option_of_records`, and the result is wrong; it _drops_ `Nones`, rather than merging them into the record fields. I'm labeling this review as "Approve" because once that's fixed and the tests pass, it will be ready to merge.

Oh, I didn't add any tests for different `axis` values. Was that straightforward/formulaic? If it follows a pattern set by the many other functions that use `axis`, then I don't think we need to explicitly test it.

Darn it, I just convinced myself that I really do need to test it. One moment...

Okay, tested it.

File: src/awkward/operations/ak_merge_option_of_records.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            high-level.
+
+    Simplifies options of records, e.g.
+
+        >>> array = ak.Array([None, {"a": 1}, {"a": 2}])
+
+    into records of options, i.e.
+
+        >>> ak.merge_option_of_records(array)
+        <Array [{a: None}, {a: 1}] type='2 * {a: ?int64}'>
```
Comment (agoose77) [inline src/awkward/operations/ak_merge_option_of_records.py]:
```suggestion
        <Array [{a: None}, {a: 1}, {a: 2}] type='3 * {a: ?int64}'>
```

**Pull request merged**